### PR TITLE
Deploy with Concourse

### PR DESCRIPTION
**Pull request by cybojenix**
Our task runner has been loaded up with [aws-cli](https://github.com/alphagov/govwifi-concourse-runner/commit/7c1c66a291771cf1c0e6790f50f2f9656b9aaccc)

Flow is, on master branch:
- Run tests + linting
- On tests + lint success success:
   - push staging image
   - run staging migrations
   - deploy staging
   - on deploy staging success:
     - on manual input: 
       - push production image
       - run production migrations
       - deploy production

Testing + linting is the same as the PR pipeline, and amounts to boilerplate around `make test lint`.

Pushing the image utilises the concourse docker-image resource to build + push the ECR.

Migrations are run using a variety of helper functions around the `aws-cli` tool to gather launch type + networking information (required due to Fargate usage).

Deploys use the `aws-cli` tool to launch new task definitions

much of the history is located in https://github.com/alphagov/govwifi-admin/pull/610

### CODE REVIEW

File: ci/tasks/scripts/aws-helpers.sh [hunk 1]
```diff
@@ -0,0 +1,100 @@
...
+        --force-new-deployment
+}
+
+function ecs_deploy() {
+  local cluster_name="${1}"
+  local service_name="${2}"
+  local regions="${3:-}"
+
+  if [[ "$regions" != '' ]]; then
+    for region in ${regions//,/ }; do
```
Comment (cybojenix) [inline ci/tasks/scripts/aws-helpers.sh]: not strictly needed yet.
Was reading the auth api deployment as that's multi-regional

File: ci/pipelines/deploy.yml [hunk 2]
```diff
@@ -0,0 +1,162 @@
+resources:
+  - name: govwifi-admin
+    type: git
+    source:
+      uri: "https://github.com/alphagov/govwifi-admin.git"
+      branch: "ci-concourse-deployment"
+
+  - name: ecr
+    # This needs to be docker-image, until registry-image supports ECR
```
Comment (emileswarts) [inline ci/pipelines/deploy.yml]: Still not entirely clear what this means.  ECR isn't supported?

File: ci/pipelines/deploy.yml [hunk 3]
```diff
@@ -0,0 +1,162 @@
...
+      - task: deploy staging
+        image: runner
+        file: govwifi-admin/ci/tasks/deploy.yml
+        params:
+          AWS_ACCESS_KEY_ID: ((deploy-access-key-id))
+          AWS_DEFAULT_REGION: ((deploy-region))
+          AWS_SECRET_ACCESS_KEY: ((deploy-secret-access-key))
+          DEPLOY_STAGE: staging
+
+  - name: production gate
```
Comment (emileswarts) [inline ci/pipelines/deploy.yml]: What does gate refer to here?

File: ci/pipelines/deploy.yml [hunk 4]
```diff
@@ -0,0 +1,162 @@
...
+          AWS_ACCESS_KEY_ID: ((deploy-access-key-id))
+          AWS_DEFAULT_REGION: ((deploy-region))
+          AWS_SECRET_ACCESS_KEY: ((deploy-secret-access-key))
+          DEPLOY_STAGE: production
+
+  - name: deploy production
+    plan:
+      - get: govwifi-admin
+        passed:
+        - migrate production
```
Comment (emileswarts) [inline ci/pipelines/deploy.yml]: Square brackets for consistency?

File: ci/tasks/deploy.yml [hunk 5]
```diff
@@ -0,0 +1,19 @@
...
+    repository: cybojenix/govwifi-concourse-runner
+
+inputs:
+  - name: govwifi-admin
+
+params:
+  DEPLOY_STAGE:
+  AWS_ACCESS_KEY_ID:
+  AWS_SECRET_ACCESS_KEY:
+  AWS_DEFAULT_REGION:
```
Comment (emileswarts) [inline ci/tasks/deploy.yml]: Does this make the params required?

File: ci/tasks/migrate.yml [hunk 6]
```diff
@@ -0,0 +1,19 @@
+---
+platform: linux
+
+image_resource:
+  type: registry-image
+  source:
+    repository: cybojenix/govwifi-concourse-runner
```
Comment (emileswarts) [inline ci/tasks/migrate.yml]: Should this point to the Alphagov repo?

File: ci/pipelines/deploy.yml (hunk 2 above)
Comment (cybojenix) [reply ci/pipelines/deploy.yml]: will remove the comment as it's more of a `TODO`.
Will add into trello to migrate from using `docker-image` (which is deprecated) to `registry-image` once the functionality we need (ECR support) is merged.

File: ci/tasks/scripts/deploy.sh [hunk 7]
```diff
@@ -0,0 +1,18 @@
+#!/bin/bash
+
+source govwifi-admin/ci/tasks/scripts/aws-helpers.sh
+
+
```
Comment (emileswarts) [inline ci/tasks/scripts/deploy.sh]: extra blank line, sorry nittt

File: ci/pipelines/deploy.yml (hunk 3 above)
Comment (cybojenix) [reply ci/pipelines/deploy.yml]: This is the manual step for triggering production deploys.
Can rename to `Confirm deploy to production`, which is the same as Jenkins?

File: ci/tasks/scripts/aws-helpers.sh (hunk 1 above)
Comment (emileswarts) [reply ci/tasks/scripts/aws-helpers.sh]: gotcha

File: ci/pipelines/deploy.yml (hunk 2 above)
Comment (emileswarts) [reply ci/pipelines/deploy.yml]: Right ok

File: ci/tasks/deploy.yml (hunk 5 above)
Comment (cybojenix) [reply ci/tasks/deploy.yml]: I believe they'll default to a blank string. The docs on this are a bit vague

File: ci/tasks/migrate.yml (hunk 6 above)
Comment (cybojenix) [reply ci/tasks/migrate.yml]: This relates to a docker hub repository, so no.

The image_resource should probably be removed, as we pass it via the pipeline now

File: ci/pipelines/deploy.yml (hunk 3 above)
Comment (emileswarts) [reply ci/pipelines/deploy.yml]: Yeah maybe? Gate doesn't mean much to me. Is it some Concourse specific concept?

File: ci/pipelines/deploy.yml [hunk 8]
```diff
@@ -0,0 +1,160 @@
...
+    jobs:
+      - Push Image
+  - name: staging
+    jobs:
+      - migrate staging
+      - deploy staging
+  - name: production
+    jobs:
+      - Confirm deploy to production
+      - migrate production
```
Comment (emileswarts) [inline ci/pipelines/deploy.yml]: Should we titleize all these job names?

**Comment by cybojenix** [comment]
Blocked until Monday

**Pull request merged**